spec/unit/pops/types/p_object_type_spec.rb

```diff
@@ -1136,6 +1140,282 @@ def compile(code)
+        notice($x =~ MyType)
+        notice($x =~ MyType[default])
+        notice($x =~ MyType[default, default])
+        notice($x =~ MyType[Enum['hello', 'yo']])
+        PUPPET
+      end
+
+      it 'parameters can be provided using named arguments' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'false', 'true'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => String,
+            p2 => String
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('good bye', 'cruel world')
+        notice($x =~ MyType)
+        notice($x =~ MyType[p1 => 'hello', p2 => 'cruel world'])
+        notice($x =~ MyType[p1 => 'good bye', p2 => 'cruel world'])
+        PUPPET
+      end
+
+      it 'no constraint != default constrant' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'false'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[String,Regexp,Type[Enum],Type[Pattern]], value => Enum['hello', 'good bye'] },
+            p2 => { type => Variant[String,Regexp,Type[Enum],Type[Pattern]], value => Pattern[/world/, /universe/] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('greetings', 'earthlings')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default])
+        PUPPET
+      end
+
+      it 'Undef can be used to prevent attribute setting' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['true', 'true', 'true', 'false'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]], value => Enum['hello', 'good bye'] },
+            p2 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]], value => Pattern[/world/, /universe/] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('hello', 'world')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default])
+        notice($x =~ MyType[default, default])
+        notice($x =~ MyType[default, undef])
+        PUPPET
+      end
+
+      it 'Required parameters cannot be set using default' do
+        expect{eval_and_collect_notices(<<-PUPPET, node)}.to raise_error(/No value provided for required type_parameter MyType\[p1\]/)
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+            p2 => { type => Variant[Undef,String,Regexp,Type[Enum],Type[Pattern]] },
+          },
+          attributes => {
+            p1 => String,
+            p2 => String
+          }]
+        $x = MyType('hello', 'world')
+        notice($x =~ MyType)
+        notice($x =~ MyType[default, default])
+        PUPPET
```

Comment:
There are two notices, and one check that there is an error - unclear from which one. If the first is supposed to error the second is not needed. If both are supposed to error, the second is not tested.

Conclusion: The first (all MyType instances are matched) must succeed, and this line should be removed I think.

Reply:
Agree, and there should be two tests. One that test that the first notice succeeds and one that test that the second fails.